front/components/assistant_builder/AssistantBuilderTablesModal.tsx

```diff
@@ -104,21 +94,75 @@ export default function AssistantBuilderTablesModal({
   const selectedManagedTables = nodes.contentNodes;
   const parentsById = nodes.parentsById;
 
+  const onManagedSelectionChange = useMemo(
+    () =>
+      (
+        dataSourceView: DataSourceViewType,
+        previousNodes: LightContentNode[],
+        node: LightContentNode,
+        parents: string[],
+        selected: boolean
+      ) => {
+        setInternalIds((internalIds) => {
+          const newIds = internalIds.filter((id) => id !== node.internalId);
+          const newNodes = previousNodes.filter(
+            (n) => n.internalId !== node.internalId
+          );
+          const newParentsById = Object.entries(
+            parentsById as Record<string, Set<string>>
+          ).reduce(
+            (acc, [key, value]) =>
+              key === node.internalId
+                ? acc
+                : {
+                    ...acc,
+                    [key]: value,
+                  },
+            {} as Record<string, Set<string>>
+          );
+
+          if (selected) {
+            newIds.push(node.internalId);
+            newNodes.push(node);
+            newParentsById[node.internalId] = new Set(
+              // This is to get the same structure/order in the fallback as the endpoint return, from leaf to root, including leaf.
+              [...parents, node.internalId].reverse()
+            );
+          }
+          // Optimistic update
+          const key = serializeUseDataSourceKey({
+            workspaceId: owner.sId,
+            dataSourceName: dataSourceView.dataSource.name,
+            internalIds: newIds,
+          });
+          setFallback((prev) => ({
+            ...prev,
+            [key]: {
+              contentNodes: newNodes,
+              parentsById: newParentsById,
+            },
+          }));
+          return newIds;
+        });
+      },
+    [owner.sId, parentsById, serializeUseDataSourceKey]
+  );
+
   async function save() {
-    if (!selectedDataSource || !selectedManagedTables) {
+    if (!selectedDataSourceView || !selectedManagedTables) {
       return;
     }
     setIsSaving(true);
 
     try {
       const tableIds = selectedManagedTables.map((n) =>
-        getTableIdForContentNode(n)
+        getTableIdForContentNode(selectedDataSourceView.dataSource, n)
       );
 
       const tables = await Promise.all(
         tableIds.map(async (id) => {
           const tableRes = await fetch(
-            `/api/w/${owner.sId}/data_sources/${selectedDataSource.name}/tables/${id}`
+            `/api/w/${owner.sId}/data_sources/${selectedDataSourceView.dataSource.name}/tables/${id}`
```

Comment:
Please add a comment with `// TODO(GROUPS_INFRA):  Move to data_source_views endpoint`.